pkg/storage/chunk/client/baidubce/bos_storage_client.go

```diff
@@ -0,0 +1,173 @@
+	return instrument.CollectedRequest(ctx, "Bos.PutObject", bosRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		body, err := bce.NewBodyFromSizedReader(object, -1)
+		if err != nil {
+			return err
+		}
+		_, err = b.client.BasicPutObject(b.cfg.BucketName, objectKey, body)
+		return err
+	})
+}
+
+func (b *BosObjectStorage) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, int64, error) {
+	var res *api.GetObjectResult
+	err := instrument.CollectedRequest(ctx, "Bos.GetObject", bosRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		var requestErr error
+		res, requestErr = b.client.BasicGetObject(b.cfg.BucketName, objectKey)
+		return requestErr
+	})
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "failed to get bos object")
+	}
+	var size int64
+	if res.ContentLength != 0 {
+		size = res.ContentLength
+	}
+	return res.Body, size, nil
+}
+
+func (b *BosObjectStorage) List(ctx context.Context, prefix string, delimiter string) ([]client.StorageObject, []client.StorageCommonPrefix, error) {
+	var storageObjects []client.StorageObject
+	var commonPrefixes []client.StorageCommonPrefix
+
+	err := instrument.CollectedRequest(ctx, "Bos.List", bosRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		args := new(api.ListObjectsArgs)
+		args.Prefix = prefix
+		args.Delimiter = delimiter
+		for {
+			listObjectResult, err := b.client.ListObjects(b.cfg.BucketName, args)
+			if err != nil {
+				return err
+			}
+			for _, content := range listObjectResult.Contents {
+				// LastModified format 2021-10-28T06:55:01Z
+				LastModifiedTime, err := time.Parse(time.RFC3339, content.LastModified)
+				if err != nil {
+					return err
+				}
+				storageObjects = append(storageObjects, client.StorageObject{
+					Key:        content.Key,
+					ModifiedAt: LastModifiedTime,
+				})
+			}
+			for _, commonPrefix := range listObjectResult.CommonPrefixes {
+				commonPrefixes = append(commonPrefixes, client.StorageCommonPrefix(commonPrefix.Prefix))
+			}
+			if !listObjectResult.IsTruncated {
+				break
+			}
+			args.Prefix = listObjectResult.Prefix
+			args.Marker = listObjectResult.NextMarker
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	return storageObjects, commonPrefixes, nil
+}
+
+func (b *BosObjectStorage) DeleteObject(ctx context.Context, objectKey string) error {
+	return instrument.CollectedRequest(ctx, "Bos.DeleteObject", bosRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		err := b.client.DeleteObject(b.cfg.BucketName, objectKey)
+		return err
+	})
+}
+
+func (b *BosObjectStorage) IsObjectNotFoundErr(err error) bool {
+	// ref: https://cloud.baidu.com/doc/BOS/s/rjwvyrwye
+	switch realErr := err.(type) {
+	case *bce.BceClientError:
+		return false
+	case *bce.BceServiceError:
+		if realErr.Code == ObjectNotFoundErr {
+			return true
+		}
```

Comment:
Will this work? I'm not understanding the link you have above the `ObjectNotFoundErr` definition, and how will you compare the string `ObjectNotFoundErr` to an error code?

Reply:
yes.It works fine.We can use this error code to represent that the storage object does not exist.I test is no problem.
This page describe it . https://intl.cloud.baidu.com/doc/BOS/s/Ajwvysfpl-en
![image](https://user-images.githubusercontent.com/17517365/164059435-48b81047-3582-4910-bc0c-d4179475c085.png)

Reply:
Okay, can you please include a comment to this [link](https://github.com/baidubce/bce-sdk-go/blob/1e5bfbecf07c6ed5d97a0090a9faee7d89466239/bce/error.go#L47-L53 ) or something similar that shows the struct definition for `BceServiceError`. I was confused because by default I expect something called `error.Code` to be an integer type.

Reply:
Yes, I agree with what you said, I think it would be better to return integer error codes here as well. 
Now I added comments pointing to its source code.